### PR TITLE
Refactor bucket sort

### PR DESCRIPTION
**Pull request by MarinPostma**
draft, experiments refactoring bucket_sort and query_builder

### CODE REVIEW

File: meilisearch-http/src/helpers/meilisearch.rs [hunk 1]
```diff
@@ -154,8 +154,13 @@ impl<'a> SearchBuilder<'a> {
...
-        query_builder.set_facets(self.facets);
+        if let Some(facet_filter) = self.facet_filter {
+            query_builder.with_facet_filter(reader, facet_filter)?;
+        }
+
+        if let Some(facets) = self.facets {
+            query_builder.with_facets_count(reader, facets)?;
+        }
 
         let start = Instant::now();
```
Comment (MarinPostma) [inline meilisearch-http/src/helpers/meilisearch.rs:165]: this should probably be before adding the facets

**Comment by Kerollmops** [comment]
Please start your branch from #691, I will merge it before you and it will reduce the amoun t of rebases to do so.

File: meilisearch-core/src/query_builder.rs [hunk 2]
```diff
@@ -88,97 +138,25 @@ impl<'c, 'f, 'd, 'i> QueryBuilder<'c, 'f, 'd, 'i> {
...
-                self.index.postings_lists,
-                self.index.documents_fields_counts,
-                self.index.synonyms,
-                self.index.prefix_documents_cache,
-                self.index.prefix_postings_lists_cache,
-            ),
+        if self.distinct.is_some() {
+            bucket_sort_with_distinct(reader, query, range, self)
+        } else {
+            bucket_sort(reader, query, range, self)
```
Comment (MarinPostma) [inline meilisearch-core/src/query_builder.rs:153]: It seems to me that bucket sort is part query_builder, given its extensive use of self...

File: meilisearch-core/src/query_builder.rs [hunk 3]
```diff
@@ -88,97 +138,25 @@ impl<'c, 'f, 'd, 'i> QueryBuilder<'c, 'f, 'd, 'i> {
...
-                Some(facet_count_map)
+        // When there is no candidate document in the facet filter, we don't perform the search,
+        // as the result will alway be empty.
+        if let Some(ref facet_filter) = self.facet_filter {
+            if facet_filter.is_empty() {
+                return Ok(SortResult::default())
             }
-            None => None,
-        };
+        }
```
Comment (MarinPostma) [inline meilisearch-core/src/query_builder.rs:148]: I'm not *super* happy with this, but this is a trade-off that's required for performance.

**Pull request closed without merging**